docs/architecture/psa-driver-interface.md

```diff
@@ -0,0 +1,561 @@
+PSA Cryptoprocessor Driver Interface
+====================================
+
+This document describes an interface for cryptoprocessor drivers in the PSA cryptography API. It describes the design of the PSA interface as well as the reference implementation in Mbed TLS and provides instructions and guidelines to driver writers.
+
+**This is work in progress**. This document is still incomplete and **may change or may be abandoned at any time**. The interface is not fully implemented in Mbed TLS yet and is disabled by default; you can enable the experimental work in progress by setting `MBEDTLS_PSA_CRYPTO_DRIVERS` in the compile-time configuration.
+
+Time-stamp: "2020/05/26 21:44:21 GMT"
+
+## Introduction
+
+### Purpose of the driver interface
+
+The PSA Cryptography API defines an interface that allows applications to perform cryptographic operations in a uniform way regardless of how the operations are performed. Under the hood, different keys may be processed in different hardware or in different logical partitions, and different algorithms may involve different hardware or software components.
+
+The driver interface allows implementations of the PSA Crypytography API to be built compositionally. An implementation of the PSA Cryptography API is composed of a **core** and zero or more **drivers**. The core handles key management, enforces key usage policies, and dispatches cryptographic operations either to the applicable driver or to built-in code.
+
+Functions in the PSA Cryptography API invoke functions in the core. Code from the core calls drivers as described in the present document.
+
+### Types of drivers
+
+The PSA Cryptography driver interface supports two types of cryptoprocessors, and accordingly two types of drivers.
+
+* **Transparent** drivers implement cryptographic operations on keys that are provided in cleartext at the beginning of each operation. They are typically used for hardware **accelerators**. When a transparent driver is available for a particular combination of parameters (cryptographic algorithm, key type and size, etc.), it is used instead of the default software implementation. Transparent drivers can also be pure software implementations that are distributed as plug-ins to a PSA Crypto implementation.
```

Comment:
Does this mean that all of mbedTLS' current software-based cryptography implementations will be migrated to look like a transparent driver?

Reply:
This is currently left open. We aren't currently planning to do it in Mbed TLS, but we've had interest for a PSA crypto layer that contains only key storage, policy enforcement and dispatch but no built-in cryptographic mechanism.

Part of the reason why we're reluctant to switch is that it would impose a dependency on the same tooling when the application writer decides what to include. We'd like to make it possible for a platform vendor to distribute a `psa_crypto_drivers_generated.c` to plug into a given version of Mbed TLS, and for an application writer to distribute a `mbedtls/config.h` (generated automatically or manually), without the platform vendor and the application writer having to coordinate. If everything is a plug-in, then the inclusion of plug-ins has to depend on the application's wishes, which intertwines the dependencies.

Reply:
If `psa_crypto_drivers_generated.c` is generated in such a way that the core implementations are always there, but behind the relevant defines, then this can still happen? A platform vendor ships an autogenerated `psa_crypto_drivers_generated.c` which includes fallback to core software implementations behind define guards, and the file gets compiled by the application writer who can set/remove the defines from config.h based on application needs.

Reply:
That would be an implementation choice: include all your crypto as a transparent driver which follows this specification and gets assembled last, or include all your crypto through a proprietary mechanism.